Add golang backend support and code generation

gustafj opened the pull request:
Hi, this patch set contains backend support and code generation output for the Go language [golang](https://golang.org/) through the `-g` option.
The golang code generation implements the golang BinaryMarshaller interface in order to be as idiomatic as possible.

The patch set incorporates and is dependent on the changes that are part of the pull requests #217, #218, #219. Thus they should be merged before this pull request is considered.

Co-authored with @TacoVox

TacoVox commented:
Hi @ashuang!
As you can see, this is quite some large PR and we know that it is a lot to review.
Nevertheless, we would extremely appreciate that this is considered to be in master of this repository.

We are actually using this code in production already and it works just fine and as expected thus far. If we encounter any further issues, we will just amend them to this PR or - in case this gets closed before - open new ones.

Needless to say, we will clean this up and squash the two commits that we added lately before this should go in.

tprk77 commented:
The main problem I see with this PR is that it doesn't integrate the build. In my opinion there needs to be at least some minimal integration for this to be a complete language binding. 

I have to admit that I'm not to familiar with Go, but it appears there's no CMake support, unfortunately. But I wonder if you could do something simple like a custom command to invoke the usual Go build tools.

Maybe you could explain how you use this Go binding in production? Like where the file is installed, and how it eventually gets built? How do Go projects consume the library?

TacoVox commented:
Hi @tprk77!
> In my opinion there needs to be at least some minimal integration for this to be a complete language binding.

As far as I am concerned this is not necessary. But let me clarify by answering your questions.

> [...] it appears there's no CMake support, unfortunately.

Of course it would be possible to build Go code with CMake. But it is not really necessary. Especially in this case here.

What we have here are basically Go bindings to the C library itself. Just as you did with the Python parts. This is possible because of a handy thing called Cgo that allows for the interconnection of Go code with C code.

> Maybe you could explain how you use this Go binding in production? Like where the file is installed, and how it eventually gets built? How do Go projects consume the library?

Go comes with a set of handy tools. One of them is `go get` which downloads packages for you. So in this case you would run a simple `go get github.com/lcm-proj/lcm/lcm-go/lcm` to download the Go library into your `$GOPATH/src/github.com/lcm-proj/lcm/lcm-go/lcm`.

The user would only need to make sure that the lcm library is properly installed (the C parts) on the system (so e.g. by using `apt install liblcm-dev` or by building and installing by hand).

Finally, the library bindings are ready to use by simply importing `github.com/lcm-proj/lcm/lcm-go/lcm`. Building and linking against the in C written LCM library is then automatically taken care of by the go compiler. It is extremely easy to use. Literally nothing you can do wrong :sunglasses: 

I hope this helps?

TacoVox commented:
Hello @tprk77.
Unfortunately I have to strongly disagree. :sunglasses:

> I think it could be helpful to some users to copy lcm.go to somewhere like `/usr/local/share/gocode/src/github.com/lcm-proj/lcm/`.

I don't see any case where this would be necessary.

Usually `$GOPATH` is not set to `/usr/local/share/gocode/src/`. In fact, I am working on a Ubuntu workstation at this moment, and `$GOPTAH` was not set to this location per default. It was set to `$HOME/go`.
When you now go ahead and create a new project, you would usually create it in `$GOPATH/src/yourproject`. Obviously this should not be `/usr/local/share/gocode/src` for development projects (rights issues, etc).

For more advance projects you would actually not use `go get`, but instead a more advanced dependency manager, such as `go dep`, which then would put the source code in a directory called `vendor` in your project folder (e.g. `$GOPATH/src/yourproject/vendor/github.com/lcm-proj/lcm`).

I hope this makes it a bit more clearer. :)

ashuang commented:
Now that the dependent PRs are in, I think we should be able to merge this one in once conflicts are resolved.

TacoVox commented:
Hey @ashuang.

That sounds great. We will add/fix some things next week (e.g. tutorial, and some cleanup) and then this is ready to get in :sunflower: 

We will keep you posted.

gustafj commented:
Hi @ashuang, we rebased the pull request, and thus fixed the conflicts so this should now merge without any issues.
We have also updated the Go backend to follow a more idiomatic way of writing Go compared to the first version posted.
The current commits in this pull request contains backend support, lcm-gen support (using `-g` and `--go*`, tests (both Go specific unit test & the 6 test cases described in the `test/README`), readme updates and tutorial examples.

gustafj commented on lcm-cmake/lcmUtilities.cmake:
```diff
@@ -7,7 +7,9 @@
+#                  [GO_SOURCES <VARIABLE_NAME>]
 #                  [DESTINATION <PATH>]
+#                  [GO_DESTINATION <PATH>]
 #                  <FILE> [<FILE>...])
```
Due to the way Go uses `$GOPATH` (implicitly needs a `src`) we need to have the Go generated files inside a `src` sub directory.
This was the simplest way to achieve that.

gustafj commented on test/README:
```diff
@@ -57,22 +57,22 @@ follows:
   .orientation[2] = n + 1
   .orientation[3] = n + 1
   .num_ranges     = n + 1
-  .ranges[i]      = i         (i = 0 ... n-1)
```
Since `.num_ranges` is `n + 1` this needs to run to `n` and not `n - 1` as previously written.
This is also what is received in the client.

gustafj commented on test/README:
```diff
@@ -192,7 +192,7 @@ n from 0 to 999:
   .another.val               = n
 
 The server should reply to each message with an lcmtest2.cross_package_t message on
-channel "test_lcmtest2_cross_package_t".  The reply should be populated as
```
Replies are sent on the `_reply channel`

gustafj commented on test/README:
```diff
@@ -166,7 +166,7 @@ The server should reply to each message with an lcmtest.multidim_array_t
 message on channel "test_lcmtest_multidim_array_t_reply".  The reply should be
 similar to the client message, but with as if the client was on message (n+1).
 
-5. cross_package_t
```
This is actually test no 6

gustafj commented on test/README:
```diff
@@ -57,22 +57,22 @@ follows:
   .enabled        = (n+1) % 2
 
 
-3. example_list_t
```
The tests are actually not using `example_list_t` but `primitives_list_t` and there is no such type in the `test/types` directory.

TacoVox commented:
Hi @ashuang 

Thanks for your openess to those additions! Really appreciate it :)
